modules/Core/classes/Avatars/MinecraftAvatarSource.php

```diff
@@ -0,0 +1,135 @@
+            self::$_active_source = self::$_sources[$source];
+        }
+
+        return self::$_active_source;
+    }
+
+    /**
+     * Get all registered sources.
+     *
+     * @return MinecraftAvatarSourceBase[]
+     */
+    private static function getAllSources(): iterable {
+        return self::$_sources;
+    }
+
+    /**
+     * Get default perspective to pass to the active avatar source.
+     *
+     * @return string Perspective.
+     */
+    private static function getPerspective(): string {
+        return Settings::get('minecraft_avatar_perspective', 'face');
+    }
+
+    /**
+     * Get raw url of active avatar source with placeholders.
+     *
+     * @return string URL with placeholders.
+     */
+    public static function getUrlToFormat(): string {
+        return self::getActiveSource()->getUrlToFormat(self::getPerspective());
+    }
+
+    /**
+     * Register avatar source.
+     *
+     * @param MinecraftAvatarSourceBase $source Instance of avatar source to register.
+     */
+    public static function registerSource(MinecraftAvatarSourceBase $source): void {
+        self::$_sources[$source->getSafeName()] = $source;
+    }
+
+    /**
+     * Get the names and base urls of all the registered avatar sources for displaying.
+     * Used for showing list of sources in staffcp.
+     *
+     * @return array<string, string> List of names.
+     */
+    public static function getAllSourceNames(): array {
+        $names = [];
+
+        foreach (self::getAllSources() as $source) {
+            $names[$source->getSafeName()] = rtrim($source->getBaseUrl(), '/') . ' ' . ($source->supportsUsernames() ? '(supports usernames)' : '');
+        }
+
+        return $names;
+    }
+
+    /**
+     * Get key value array of all registered sources and their available perspectives.
+     * Used for autoupdating dropdown selector in staffcp.
+     *
+     * @return array<string, array<string>> Array of source => [] perspectives.
+     */
+    public static function getAllPerspectives(): array {
+        $perspectives = [];
+
+        foreach (self::getAllSources() as $source) {
+            foreach ($source->getPerspectives() as $perspective) {
+                $perspectives[$source->getSafeName()][] = $perspective;
+            }
+        }
+
+        return $perspectives;
+    }
+
+    protected function get(User $user): ?string {
+        $minecraft_integration = $user->getIntegration('Minecraft');
+        if ($minecraft_integration !== null) {
+            $identifier = $minecraft_integration->data()->identifier;
+        } else {
+            $identifier = $user->data()->username;
+            // Fallback to steve avatar if they have an invalid username
+            // TODO: what is this regex?
+            if (preg_match('#[^][_A-Za-z0-9]#', $identifier)) {
+                $identifier = 'Steve';
+            }
+        }
+
+        return self::getAvatarFromIdentifier($identifier, $this->_size);
+    }
```

Comment:
We should probably check if the uuid is an offline one and use usernames instead if this is the case. This should fix the fact that when force premium accounts is disabled, people only receive steve skins. Nameless generates an offline uuid when force premium accounts is enabled which would still make this use that offline uuid for retrieving accounts